crates/accelerate/src/uc_gate.rs

```diff
@@ -0,0 +1,160 @@
+
+    // If d is not equal to diag(i,-i), then we put it into this "standard" form
+    // (see eq. (13) in https://arxiv.org/pdf/quant-ph/0410066.pdf) by interchanging
+    // the eigenvalues and eigenvectors
+    if (diag[0] + Complex64::new(0., 1.)).abs() < EPS {
+        diag = diag.slice(s![..;-1]).to_owned();
+        u = u.slice(s![.., ..;-1]).to_owned();
+    }
+    diag.mapv_inplace(|x| x.sqrt());
+    let d = Array2::from_diag(&diag);
+    let v = d
+        .dot(&u.mapv(|x| x.conj()).t())
+        .dot(&r.mapv(|x| x.conj()).t())
+        .dot(&b);
+    [v, u, r]
+}
+
+#[pyfunction]
+pub fn dec_ucg_help(
+    py: Python,
+    sq_gates: Vec<PyReadonlyArray2<Complex64>>,
+    num_qubits: u32,
+) -> (Vec<PyObject>, PyObject) {
+    let mut single_qubit_gates: Vec<Array2<Complex64>> = sq_gates
+        .into_iter()
+        .map(|x| x.as_array().to_owned())
+        .collect();
+    let mut diag: Array1<Complex64> = Array1::ones(2_usize.pow(num_qubits));
+    let num_controls = num_qubits - 1;
+    for dec_step in 0..num_controls {
+        let num_ucgs = 2_u32.pow(dec_step);
+        // The decomposition works recursively and the followign loop goes over the different
+        // UCGates that arise in the decomposition
+        for ucg_index in 0..num_ucgs {
+            let len_ucg = 2_u32.pow(num_controls - dec_step);
+            for i in 0..len_ucg / 2 {
+                let shift = ucg_index * len_ucg;
+                let a = single_qubit_gates[(shift + i) as usize].view();
+                let b = single_qubit_gates[(shift + len_ucg / 2 + i) as usize].view();
+                // Apply the decomposition for UCGates given in equation (3) in
+                // https://arxiv.org/pdf/quant-ph/0410066.pdf
+                // to demultiplex one control of all the num_ucgs uniformly-controlled gates
+                // with log2(len_ucg) uniform controls
+                let [v, u, r] = demultiplex_single_uc(a, b);
+                // replace the single-qubit gates with v,u (the already existing ones
+                // are not needed any more)
+                single_qubit_gates[(shift + i) as usize] = v;
+                single_qubit_gates[(shift + len_ucg / 2 + i) as usize] = u;
+                // Now we decompose the gates D as described in Figure 4 in
+                // https://arxiv.org/pdf/quant-ph/0410066.pdf and merge some of the gates
+                // into the UCGates and the diagonal at the end of the circuit
+
+                // Remark: The Rz(pi/2) rotation acting on the target qubit and the Hadamard
+                // gates arising in the decomposition of D are ignored for the moment (they will
+                // be added together with the C-NOT gates at the end of the decomposition
+                // (in the method dec_ucg()))
+                let rz_11 = (-Complex64::new(0., 0.5 * PI2)).exp();
+                let rz_00 = Complex64::new(0., 0.5 * PI2).exp();
```

Comment:
Lol, oh I didn't even see that I just was in mechanical porting mode. Do you think it's worth making that explicit here, or just leave it for LLVM to optimize?

Reply:
I'm fine leaving all the little things like this - larger-scale _real_ performance improvements in this would come from a more complete refactor of the implementation as a whole, and as you say, the compiler can probably constant-fold this anyway.